tools/syz-trace2syz/parser/intermediate_types.go

```diff
@@ -0,0 +1,531 @@
+	CallName string
+	Args     []IrType
+	Pid      int64
+	Ret      int64
+	Cover    []uint64
+	Paused   bool
+	Resumed  bool
+}
+
+// NewSyscall - constructor
+func NewSyscall(pid int64, name string, args []IrType, ret int64, paused, resumed bool) (sys *Syscall) {
+	return &Syscall{
+		CallName: name,
+		Args:     args,
+		Pid:      pid,
+		Ret:      ret,
+		Paused:   paused,
+		Resumed:  resumed,
+	}
+}
+
+// String
+func (s *Syscall) String() string {
+	buf := new(bytes.Buffer)
+
+	fmt.Fprintf(buf, "Pid: -%v-", s.Pid)
+	fmt.Fprintf(buf, "Name: -%v-", s.CallName)
+	for _, typ := range s.Args {
+		buf.WriteString("-")
+		buf.WriteString(typ.String())
+		buf.WriteString("-")
+	}
+	buf.WriteString(fmt.Sprintf("-Ret: %d\n", s.Ret))
+	return buf.String()
+}
+
+// GenDefaultIrType generates a default intermediate type from our own types
+func GenDefaultIrType(syzType prog.Type) IrType {
+	switch a := syzType.(type) {
+	case *prog.StructType:
+		straceFields := make([]IrType, len(a.Fields))
+		for i := 0; i < len(straceFields); i++ {
+			straceFields[i] = GenDefaultIrType(a.Fields[i])
+		}
+		return newGroupType(straceFields)
+	case *prog.ArrayType:
+		straceFields := make([]IrType, 1)
+		straceFields[0] = GenDefaultIrType(a.Type)
+		return newGroupType(straceFields)
+	case *prog.ConstType, *prog.ProcType, *prog.LenType, *prog.FlagsType, *prog.IntType:
+		return NewIntsType([]int64{0})
+	case *prog.PtrType:
+		return NewPointerType(0, GenDefaultIrType(a.Type))
+	case *prog.UnionType:
+		return GenDefaultIrType(a.Fields[0])
+	default:
+		log.Fatalf("Unsupported syz type for generating default strace type: %s", syzType.Name())
+	}
+	return nil
+}
+
+// Call Represents arguments that are expanded by strace into calls
+// E.g. inet_addr("127.0.0.1")
+type Call struct {
+	expressionCommon
+	CallName string
+	Args     []IrType
+}
+
+func newCallType(name string, args []IrType) *Call {
+	return &Call{CallName: name, Args: args}
+}
+
+// String implements IrType String()
+func (c *Call) String() string {
+	buf := new(bytes.Buffer)
+	buf.WriteString("Name: " + c.CallName + "\n")
+	for _, arg := range c.Args {
+		buf.WriteString("Arg: " + arg.Name() + "\n")
+	}
+	return buf.String()
+}
+
+// Eval implements Expression's Eval()
+func (c *Call) Eval(target *prog.Target) uint64 {
+	return EvalCalls(target, c)
+}
+
+// GroupType contains arrays and structs
+type GroupType struct {
+	Elems []IrType
+	Len   int
+}
+
+func newGroupType(elems []IrType) (typ *GroupType) {
+	return &GroupType{Elems: elems, Len: len(elems)}
+}
+
+// Name implements IrType Name()
+func (a *GroupType) Name() string {
+	return groupTypeName
+}
+
+// String implements IrType String()
+func (a *GroupType) String() string {
+	var buf bytes.Buffer
+
+	buf.WriteString("[")
+	for _, elem := range a.Elems {
+		buf.WriteString(elem.String())
+		buf.WriteString(",")
+	}
+	buf.WriteString("]")
+	return buf.String()
+}
+
+// Field of a struct e.g. name = "Shankara"
+type Field struct {
+	Key string
```

Comment:
Key is not used anywhere. Which means that we can remove Field and just use Val expression instead of it. Which should get rid of all this code in proggen:
```
$ grep -A 1 "*parser.Field" proggen/*.go
proggen/generate_unions.go:			case *parser.Field:
proggen/generate_unions.go-				curArg := ctx.CurrentStraceArg
--
proggen/generate_unions.go:	case *parser.Field:
proggen/generate_unions.go-		idx = 2
--
proggen/proggen.go:	case *parser.Field:
proggen/proggen.go-		return genArray(syzType, a.Val, ctx)
--
proggen/proggen.go:	case *parser.Field:
proggen/proggen.go-		return genArgs(syzType, a.Val, ctx)
--
proggen/proggen.go:	case *parser.Field:
proggen/proggen.go-		return genUnionArg(syzType, strType.Val, ctx)
--
proggen/proggen.go:		case *parser.Field:
proggen/proggen.go-			return genBuffer(syzType, a.Val, ctx)
--
proggen/proggen.go:	case *parser.Field:
proggen/proggen.go-		return genArgs(syzType, a.Val, ctx)
--
proggen/proggen.go:	case *parser.Field:
proggen/proggen.go-		return genPtr(syzType, a.Val, ctx)
--
proggen/proggen.go:	case *parser.Field:
proggen/proggen.go-		// We have an argument of the form sin_port=IntType(0)
--
proggen/proggen.go:	case *parser.Field:
proggen/proggen.go-		return genResource(syzType, a.Val, ctx)
--
proggen/proggen.go:	case *parser.Field:
proggen/proggen.go-		return genArgs(syzType, a.Val, ctx)
```

Reply:
Yeah I agree. I think I mentioned that in a comment above.